### ISSUE TITLE
Enable tree shaking of platform-specific code in AOT mode.

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#31969]
Based on the converstation with Flutter users it seems that Flutter code it is common to guard large chunks of code behind runtime checks that boil down to:

```dart
if (Platform.isAndroid) {
  // ...
}
```

(actual flutter code might be using a layer on top: see https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/platform.dart)

Here's a list of subtasks:

* [x] #50472
* [x] #50473
* [ ] #50474
* [ ] Eventually we may add annotation of platform-specific code + warning mechanism if not guarded by platform checks to Dart language tools (CFE, Analyzer) 
* [ ] Investigate whether above tree shaking mechanisms are sufficient for flutter plugin code.

/cc @lukef @mit-mit

**Comment by mkustermann** [comment]
Cross-linking some other issues

  * https://github.com/flutter/flutter/issues/14233
  * https://github.com/dart-lang/sdk/issues/35705
  * https://github.com/dart-lang/language/issues/1889

**Comment by mkustermann** [comment]
The outcome of our offline discussion on Sept 29 was that we'll define a set of compiler-recognizable patterns for code that is dependent on platform/os/arch/... and then make our AOT compiler recognize those patterns to perform tree shaking on the kernel level.

@lrhn Have you thought about the set of patterns we should make our compiler support?

**Comment by lrhn** [comment]
My idea is fairly simple: https://gist.github.com/lrhn/2a3b06a8253fddaf61297cb604c6777f

Basically, recognize `isLinux` and `operatingSystem == "linux"`, and combinations of those like `isLinux || isWindows`.

**Comment by mkustermann** [comment]
> My idea is fairly simple: https://gist.github.com/lrhn/2a3b06a8253fddaf61297cb604c6777f
> 
> Basically, recognize isLinux and operatingSystem == "linux", and combinations of those like isLinux || isWindows.,

Thanks for writing it up, @lrhn !

What would be the code patterns we support on the usage site of those? (e.g. `if (isLinux)`, caching `isLinux` in a local variable, using `isLinux` in other getters, more complicated control flow, ...)

Flutter is having it's own platform detection mechanism atm (it is exposing a [defaultTargetPlatform](https://github.com/flutter/flutter/blob/d9a2229babbc7e28cabd/packages/flutter/lib/src/foundation/platform.dart#L39) getter that is based on [_platform_io.dart](https://github.com/flutter/flutter/blob/d9a2229babbc7e28cabd/packages/flutter/lib/src/foundation/_platform_io.dart#L12) which is currently even overridable at runtime by setting OS environment variable `FLUTTER_TEST` if assertions are enabled).
```dart
platform.TargetPlatform get defaultTargetPlatform {
  platform.TargetPlatform? result;
  if (Platform.isAndroid) {
    result = platform.TargetPlatform.android;
  } else if (Platform.isIOS) {
    ...
  }
  assert(() {
    if (Platform.environment.containsKey('FLUTTER_TEST')) {
      result = platform.TargetPlatform.android;
    }
    return true;
  }());
  if (platform.debugDefaultTargetPlatformOverride != null) {
    result = platform.debugDefaultTargetPlatformOverride;
  }
  if (result == null) throw ...;
  return result!;
}
```
Would we want to make our compiler recognize the above code to return a constant? Would we want to rewrite this implementation in flutter?

This flutter `defaultTargetPlatform`  is used in practice to guard code, for example the flutter firebase plugin uses it [here](https://github.com/firebase/flutterfire/blob/8a0caa05d5abf6fef5bf0e654654dcd0b6ec874a/packages/firebase_messaging/firebase_messaging_platform_interface/lib/src/method_channel/method_channel_messaging.dart#L172)

**Comment by lrhn** [comment]
I'd prefer to not make things too complicated, so only direct occurrences of `isLinux` inside the condition expression itself.
Remembering inferences from a boolean stored in a variable is possible (we've proven that for promotion), but it's also much more complicated and a little harder to read. The `isX` variables are short and directly useful.

I definitely don't want `if (myPlatformTest()) { ... }` to work, where `myPlatformTest() => isLinux || isAndroid;`. (If we ever make promoting methods work for type promotions, I'm willing to look at it again.)

As for more complicated control flow ... only if reasonable. (That would be recognizing escaping control flow.)
It's safe to not start with it. The warnings apply whenever you use a platform specific method, unless we can deduce that you can only be on a supported platform (can't rule out that you are on an unsupported platform). Making our analysis better in the future will not introduce more errors, only fewer.

I'd probably make the compiler just check for the `operatingSystem == "linux"` (and `"linux" == operatingSystem`), and force-inline any getter starting with `is` from the `package:platform/platform.dart` library. Then we can add more `isGroupOfPlatforms => isFoo || isBar || isBaz;` getters if necessary without needing to add extra support.

We can *choose* to accept `assert(isLinux);` too, for warnings, even if we can't use it for optimizations when asserts are not enabled.


Flutter could rewrite their target platform code to use our getters. It won't make the result *constant*, but it might be known at compile-time anyway, if the compiler knows the value of, say, `isAndroid`. That's a valid optimization.
Since they support overriding the *actual* runtime platform, it's different from what we do. The Dart compilers can safely optimize away non-platform code based on the values of the `isX` getters when it knows the value at runtime. The Flutter code above uses a non-`const` check for `Platform.environment`, so we can't do compile-time optimization based on the result of `defaultTargetPlatform`.

The goal here is two things:
* Primary: Get warnings if using platform-specific code when not definitely on that platform. 
    * This is a _static_ check.
    * This is best effort.
    * We can decide on any number of ways to decide the current platform. Better analysis only removes warnings.
    * We can even use `assert(isLinux)` as a sign that the following code is definitely running on Linux, even if it's not *safe*.
* Secondary: Optimize AoT tree-shaking by removing any code guarded by check for a platform other than the one being compiled for. 
    * This too must be best effort.
    * It must be dynamically safe, so we can't rely on `assert(isLinux)` for truth in production, which is where you really want the tree-shaking.
    * It must not change program behavior, so we can't allow you to mock the current platform dynamically, it must at least go through the compiler.

If used correctly, all code that doesn't run on the current platform should be tree-shakable, because any use of it is definitely inside other code which can be tree-shakable.

**Comment by mkustermann** [comment]
> Flutter could rewrite their target platform code to use our getters. It won't make the result constant, ...

We **have to** be able to tree shake code that is guarded by platform checks. If flutter provides a blessed API that they recommend to users, users will use it, so we have to be able to tree shake based on that API.

(Seemingly this was originally added in https://github.com/flutter/flutter/pull/33663)

**Comment by mkustermann** [comment]
After some more offline discussion, I think we'll end up with a few action items:

* [ ] Make `flutter build` / `dart compile exe` / ... tell VM's global analysis pass which operating system we compile for
    => @sstrickl will work on this
* [ ] Add mechanism to instruct VM's global analysis pass to try to const evaluate certain parts of a program before global analysis runs (possibly by recognizing e.g. `@pramga('vm:try-evaluate-as-constant')`) - it will support subset of language and recognizes calls to `dart:io`s platform getters as constant.
    => @sstrickl will work on this
* [ ] Make `package:platform` APIs that wrap `dart:io` / `dart:html` APIs and take advantage of above mechanism
    => @lrhn @mit-mit will do this.
* [ ] Update Flutter documentation of [defaultTargetPlatform](https://github.com/flutter/flutter/blob/d9a2229babbc7e28cabd/packages/flutter/lib/src/foundation/platform.dart#L39) to recommend using `package:platform` if one wants tree shaking.
    => @lrhn @mit-mit will do this after `package:platform` is ready.
* [ ] Eventually we may add annotation of platform-specific code + warning mechanism if not guarded by platform checks to Dart language tools (CFE, Analyzer) 
* [ ] Investigate whether above tree shaking mechanisms are sufficient for flutter plugin code.

When using the annotation based approach, the VM is unblocked and can implement the first two items.

**Comment by sstrickl** [comment]
Since we now have the `"vm:platform-const[-if]"` pragmas, `dart:io` and `debugTargetPlatform` have been updated to use them appropriately for platform-related fields and checks, and the pragma is also available to users to build their own custom platform checks based on what's provided by `dart:io` and `debugTargetPlatform` (e.g., writing predicates that are combinations of the base predicates), I'm closing this out.

If we want to do static analysis that attempts to detect platform-specific code that isn't guarded by a platform check, that should likely go into a separate issue for further discussion and resource allocation.

(There is already have a separate issue for updating `dart:platform` to use the available mechanisms.)

**Comment by alex-lechner** [comment]
@sstrickl would you mind to update the flutter documentation with a description of those mentioned variables and on how to use them, please? I couldn't find this anywhere
